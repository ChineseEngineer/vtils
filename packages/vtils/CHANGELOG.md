# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [2.26.1-beta.1](https://github.com/fjc0k/vtils/compare/v2.26.1-beta.0...v2.26.1-beta.1) (2019-09-10)


### Bug Fixes

* 修复一些问题 ([d36cc62](https://github.com/fjc0k/vtils/commit/d36cc62))
