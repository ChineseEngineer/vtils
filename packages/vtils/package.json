{
  "name": "vtils",
  "version": "2.34.0",
  "description": "小巧实用的 JavaScript 工具类库。",
  "main": "lib/index.js",
  "module": "lib/index.esm.js",
  "unpkg": "lib/index.umd.js",
  "jsdelivr": "lib/index.umd.js",
  "types": "lib/index.d.ts",
  "homepage": "https://github.com/fjc0k/vtils/tree/master/packages/vtils",
  "license": "MIT",
  "sideEffects": false,
  "repository": {
    "type": "git",
    "url": "git@github.com:fjc0k/vtils.git",
    "directory": "packages/vtils"
  },
  "bugs": {
    "url": "https://github.com/fjc0k/vtils/issues"
  },
  "keywords": [
    "utils"
  ],
  "files": [
    "lib"
  ],
  "author": {
    "name": "Jay Fong",
    "email": "fjc0kb@gmail.com",
    "url": "https://github.com/fjc0k"
  },
  "scripts": {
    "buildLib": "ts-node --compiler-options '{\"module\":\"commonjs\"}' --files scripts/buildLib.ts",
    "buildDocs": "ts-node --compiler-options '{\"module\":\"commonjs\"}' --files scripts/buildDocs.ts",
    "releaseDocs": "gh-pages --dist .typedoc --dest vtils",
    "version": "yarn buildLib && yarn buildDocs && git add -A && yarn releaseDocs"
  },
  "devDependencies": {
    "@microsoft/api-extractor": "^7.4.1",
    "@tarojs/taro-h5": "^1.3.18",
    "@types/fs-extra": "^7.0.0",
    "@types/qs": "^6.5.3",
    "@types/shelljs": "^0.8.5",
    "bili": "^4.8.0",
    "fs-extra": "^8.0.1",
    "gh-pages": "^2.0.1",
    "globby": "^9.2.0",
    "miniprogram-api-typings": "^2.8.2",
    "nervjs": "^1.4.4",
    "qs": "^6.7.0",
    "rollup-plugin-typescript2": "^0.21.1",
    "shelljs": "^0.8.3",
    "ts-node": "^8.0.2",
    "typedoc": "^0.15.0-0",
    "typescript": "^3.4.5"
  }
}
