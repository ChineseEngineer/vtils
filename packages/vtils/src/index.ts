// @index(['./*', '!./_*', '!./typings.d.ts', '!./*.test.ts'], pp => `export * from '${pp.path}'`)
export * from './assign'
export * from './base64'
export * from './castArray'
export * from './chunk'
export * from './clamp'
export * from './combine'
export * from './debounce'
export * from './dedent'
export * from './defaultTo'
export * from './Disposer'
export * from './EasyStorage'
export * from './EasyValidator'
export * from './endsWith'
export * from './enhanceJest'
export * from './enhanceType'
export * from './entries'
export * from './env'
export * from './escapeRegExp'
export * from './EventBus'
export * from './fill'
export * from './flat'
export * from './flexible'
export * from './forOwn'
export * from './getType'
export * from './groupBy'
export * from './has'
export * from './ii'
export * from './includes'
export * from './indent'
export * from './inRange'
export * from './is'
export * from './keyBy'
export * from './keys'
export * from './last'
export * from './loadResource'
export * from './loop'
export * from './mapValues'
export * from './memoize'
export * from './noop'
export * from './omit'
export * from './orderBy'
export * from './pad'
export * from './parallel'
export * from './parseCSSValue'
export * from './partial'
export * from './partialBy'
export * from './pick'
export * from './placeKitten'
export * from './pluck'
export * from './randomString'
export * from './range'
export * from './remove'
export * from './removeByValue'
export * from './repeat'
export * from './result'
export * from './round'
export * from './safeGet'
export * from './sample'
export * from './sequential'
export * from './shuffle'
export * from './startsWith'
export * from './sum'
export * from './throttle'
export * from './times'
export * from './tryGet'
export * from './unique'
export * from './URI'
export * from './values'
export * from './wait'
export * from './Wechat'
