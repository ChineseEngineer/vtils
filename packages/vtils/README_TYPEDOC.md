<p align="center"><img width="200" src="https://raw.githubusercontent.com/fjc0k/vtils/master/logo.png" alt="logo"></p>

<p align="center"><a href="https://www.npmjs.com/package/vtils"><img src="https://badge.fury.io/js/vtils.svg" alt="NPM Version"></a> <a href="https://travis-ci.org/fjc0k/vtils"><img src="https://travis-ci.org/fjc0k/vtils.svg?branch=master" alt="Build Status"></a> <a href="https://codecov.io/gh/fjc0k/vtils"><img src="https://codecov.io/gh/fjc0k/vtils/branch/master/graph/badge.svg" alt="Coverage Status"></a> <img src="https://badgen.net/bundlephobia/min/vtils" alt="Size"> <img src="https://badgen.net/bundlephobia/minzip/vtils" alt="Gzip Size"> <img src="https://badgen.net/github/license/fjc0k/vtils" alt="License"></p>

<h2 align="center">小巧实用的 JavaScript 工具类库。</h2>

<p align="center">
  <a href="https://fjc0k.github.io/vtils/">https://fjc0k.github.io/vtils/vtils</a>
</p>

## 特性

- 源于日常项目实践，更实用
- 使用 TypeScript 编写，类型友好
- 支持摇树优化(Tree Shaking)，只引入使用到的工具
- 浏览器、Node、小程序多端兼容

## 说明

`vtils` 自身并不包括一些已有成熟库的工具，如时间处理、网络请求等，在此做下推荐：

- 时间处理：[dayjs](https://github.com/iamkun/dayjs)
- 网络请求：[axios](https://github.com/axios/axios)、[taro-axios](https://github.com/fjc0k/taro-axios)

## 安装

```bash
# yarn
yarn add vtils

# or, npm
npm i vtils --save
```

你也可通过 CDN 安装，然后使用全局变量 `vtils` 访问相关工具：

```html
<script src="https://cdn.jsdelivr.net/npm/vtils@2.42.0/lib/index.umd.min.js" crossorigin="anonymous"></script>

<script>
  if (vtils.inBrowser()) {
    alert('您在浏览器中...')
  }
</script>
```

## 使用

在线体验：[https://stackblitz.com/edit/vtils](https://stackblitz.com/edit/vtils)

```js
import { inBrowser, shuffle } from 'vtils'

if (inBrowser()) {
  alert('您在浏览器中...')
}

alert(shuffle([1, 2, 3, 4, 5]))
```

