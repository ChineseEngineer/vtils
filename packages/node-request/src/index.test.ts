import fs from 'fs'
import http from 'http'
import path from 'path'
import {AsyncReturnType, escapeRegExp, ii, isFunction, isPlainObject} from 'vtils'
import {NodeRequest} from '.'

let port = 4444

async function withServer({
  statusCode = 200,
  response,
}: {
  statusCode?: number,
  response: ((req: http.IncomingMessage, res: http.OutgoingMessage) => any) | Record<string, any>,
}) {
  const server = http
    .createServer(async (req, res) => {
      const actualResponse = isFunction(response)
        ? await (response as any)(req, res)
        : response
      res.writeHead(statusCode, {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
      })
      res.end(isPlainObject(actualResponse) ? JSON.stringify(actualResponse) : actualResponse)
    })
    .listen(++port)

  return Promise.resolve({
    url: `http://localhost:${port}`,
    closeServer: () => server.close(),
  })
}

describe('请求方式', () => {
  const nr = new NodeRequest()
  let server!: AsyncReturnType<typeof withServer>
  beforeAll(async () => {
    server = await withServer({
      response: async (req) => ({
        method: req.method,
      }),
    })
  })
  afterAll(() => server.closeServer())

  test('GET', async () => {
    const res = await nr.request({
      url: server.url,
      method: 'GET',
    })
    expect(res.data).toMatchSnapshot()
  })

  test('POST', async () => {
    const res = await nr.request({
      url: server.url,
      method: 'POST',
    })
    expect(res.data).toMatchSnapshot()
  })
})

describe('请求数据', () => {
  const nr = new NodeRequest()
  let server!: AsyncReturnType<typeof withServer>
  beforeAll(async () => {
    server = await withServer({
      response: req => new Promise(resolve => {
        let str = ''
        req.on('data', (chunk: Buffer) => {
          str += chunk.toString('utf8')
        })
        req.on('end', () => {
          resolve({
            url: req.url,
            method: req.method,
            contentType: req.headers['content-type'],
            contentLength: req.headers['content-length'],
            data: str,
          })
        })
      }),
    })
  })
  afterAll(() => server.closeServer())

  test('queryData', async () => {
    const res = await nr.request({
      url: server.url,
      method: 'GET',
      queryData: {x: 1, 2: 3, null: null},
    })
    expect(res.data).toMatchSnapshot()
  })

  test('formData', async () => {
    const res = await nr.request({
      url: server.url,
      method: 'POST',
      formData: {x: 1, 2: 3, null: null},
    })
    expect(res.data).toMatchSnapshot()
  })

  test('jsonData', async () => {
    const res = await nr.request({
      url: server.url,
      method: 'POST',
      jsonData: {x: 1, 2: 3, null: null},
    })
    expect(res.data).toMatchSnapshot()
  })

  test('fileData', async () => {
    const xRes = await nr.request({
      url: 'https://jsonplaceholder.typicode.com/posts',
      method: 'POST',
      jsonData: {XXXXX: 'YYYYYY'},
      responseType: 'stream',
    })
    const res = await nr.request<{ contentType: string, data: string }>({
      url: server.url,
      method: 'POST',
      queryData: {id: 100},
      formData: {x: 1, 2: 3, null: null},
      fileData: {
        jsonFile: path.join(__dirname, '../tsconfig.json'),
        jsonFile2: fs.createReadStream(path.join(__dirname, '../tsconfig.json')),
        jsonFile3: {
          file: xRes.data,
          options: {
            contentType: xRes.headers['content-type'],
            knownLength: xRes.headers['content-length'] ? Number(xRes.headers['content-length']) : undefined,
          },
        },
      },
    })
    res.data = ii(() => {
      const boundary = res.data.contentType.match(/boundary=([^"]+)/s)![1]
      res.data.contentType = res.data.contentType.replace(
        new RegExp(escapeRegExp(boundary), 'g'),
        '___boundary___',
      )
      res.data.data = res.data.data.replace(
        new RegExp(escapeRegExp(boundary), 'g'),
        '___boundary___',
      )
      return res.data
    })
    expect(res.data).toMatchSnapshot()
  })
})

describe('自定义头部', () => {
  const nr = new NodeRequest()
  let server!: AsyncReturnType<typeof withServer>
  beforeAll(async () => {
    server = await withServer({
      response: async (req) => req.headers,
    })
  })
  afterAll(() => server.closeServer())

  test('表现正常', async () => {
    const res = await nr.request({
      url: server.url,
      method: 'GET',
      headers: {
        'cookie': 'hello=1',
        'x-token': 'test_token',
        '222': '3333',
      },
    })
    expect(res.data).toMatchSnapshot()
  })
})

describe('cookie jar', () => {
  let server!: AsyncReturnType<typeof withServer>
  beforeAll(async () => {
    server = await withServer({
      response: async (req, res) => {
        res.setHeader('Set-Cookie', 'test=1')
        return {
          method: req.method,
          cookie: req.headers.cookie,
        }
      },
    })
  })
  afterAll(() => server.closeServer())

  test('没有开启 cookie jar 支持', async () => {
    const nr = new NodeRequest()
    const firstRes = await nr.request({
      url: server.url,
      method: 'GET',
    })
    expect(firstRes.data).toMatchSnapshot('首次请求 - 无 cookie')
    const secondRes = await nr.request({
      url: server.url,
      method: 'GET',
    })
    expect(secondRes.data).toMatchSnapshot('第二次请求 - 同样无 cookie')
  })

  test('开启 cookie jar 支持', async () => {
    const nr = new NodeRequest({
      cookieJar: {enabled: true},
    })
    const firstRes = await nr.request({
      url: server.url,
      method: 'GET',
    })
    expect(firstRes.data).toMatchSnapshot('首次请求 - 无 cookie')
    const secondRes = await nr.request({
      url: server.url,
      method: 'GET',
    })
    expect(secondRes.data).toMatchSnapshot('第二次请求 - 有 cookie')
  })
})

describe('静态方法', () => {
  let server!: AsyncReturnType<typeof withServer>
  beforeAll(async () => {
    server = await withServer({
      response: async (req) => {
        return {
          method: req.method,
          headers: req.headers,
        }
      },
    })
  })
  afterAll(() => server.closeServer())

  test('get', async () => {
    const res = await NodeRequest.get(server.url, {
      headers: {
        'x-get': '1',
      },
    })
    expect(res.data).toMatchSnapshot()
  })

  test('post', async () => {
    const res = await NodeRequest.post(server.url, {
      headers: {
        'x-post': '1',
      },
    })
    expect(res.data).toMatchSnapshot()
  })
})
