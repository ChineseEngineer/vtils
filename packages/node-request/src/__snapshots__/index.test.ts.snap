// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cookie jar 开启 cookie jar 支持: 第二次请求 - 有 cookie 1`] = `
Object {
  "cookie": "test=1",
  "method": "GET",
}
`;

exports[`cookie jar 开启 cookie jar 支持: 首次请求 - 无 cookie 1`] = `
Object {
  "method": "GET",
}
`;

exports[`cookie jar 没有开启 cookie jar 支持: 第二次请求 - 同样无 cookie 1`] = `
Object {
  "method": "GET",
}
`;

exports[`cookie jar 没有开启 cookie jar 支持: 首次请求 - 无 cookie 1`] = `
Object {
  "method": "GET",
}
`;

exports[`自定义头部 表现正常 1`] = `
Object {
  "222": "3333",
  "accept": "application/json, text/plain, */*",
  "connection": "close",
  "content-type": "application/x-www-form-urlencoded",
  "cookie": "hello=1",
  "host": "localhost:4447",
  "user-agent": "axios/0.19.0",
  "x-token": "test_token",
}
`;

exports[`请求数据 fileData 1`] = `
Object {
  "contentLength": "1148",
  "contentType": "multipart/form-data; boundary=___boundary___",
  "data": "--___boundary___
Content-Disposition: form-data; name=\\"2\\"

3
--___boundary___
Content-Disposition: form-data; name=\\"x\\"

1
--___boundary___
Content-Disposition: form-data; name=\\"jsonFile\\"; filename=\\"tsconfig.json\\"
Content-Type: application/json

{
  \\"extends\\": \\"../../tsconfig.json\\",
  \\"compilerOptions\\": {
    \\"target\\": \\"es6\\",
    \\"module\\": \\"commonjs\\",
    \\"outDir\\": \\"./lib\\"
  },
  \\"include\\": [
    \\"src/index.ts\\"
  ]
}

--___boundary___
Content-Disposition: form-data; name=\\"jsonFile2\\"; filename=\\"tsconfig.json\\"
Content-Type: application/json

{
  \\"extends\\": \\"../../tsconfig.json\\",
  \\"compilerOptions\\": {
    \\"target\\": \\"es6\\",
    \\"module\\": \\"commonjs\\",
    \\"outDir\\": \\"./lib\\"
  },
  \\"include\\": [
    \\"src/index.ts\\"
  ]
}

--___boundary___
Content-Disposition: form-data; name=\\"jsonFile3\\"; filename=\\"posts\\"
Content-Type: application/json; charset=utf-8

{
  \\"XXXXX\\": \\"YYYYYY\\",
  \\"id\\": 101
}
--___boundary___--
",
  "method": "POST",
  "url": "/?id=100",
}
`;

exports[`请求数据 formData 1`] = `
Object {
  "contentLength": "17",
  "contentType": "application/x-www-form-urlencoded",
  "data": "2=3&x=1&null=null",
  "method": "POST",
  "url": "/",
}
`;

exports[`请求数据 jsonData 1`] = `
Object {
  "contentLength": "25",
  "contentType": "application/json;charset=utf-8",
  "data": "{\\"2\\":3,\\"x\\":1,\\"null\\":null}",
  "method": "POST",
  "url": "/",
}
`;

exports[`请求数据 queryData 1`] = `
Object {
  "contentType": "application/x-www-form-urlencoded",
  "data": "",
  "method": "GET",
  "url": "/?2=3&x=1",
}
`;

exports[`请求方式 GET 1`] = `
Object {
  "method": "GET",
}
`;

exports[`请求方式 POST 1`] = `
Object {
  "method": "POST",
}
`;
